book/src/dev/rfcs/0003-inventory-tracking.md

```diff
@@ -0,0 +1,236 @@
+could cause the channel to fill.  However, because inventory advertisements are
+time-limited, in the sense that they're only useful before some item is fully
+distributed across the network, it's safe to handle excess entries by dropping
+them.  This behavior is provided by a [broadcast]/mpmc channel, which can be
+used in place of an mpsc channel.
+
+[mpsc]: https://docs.rs/tokio/0.2.22/tokio/sync/mpsc/index.html
+[broadcast]: https://docs.rs/tokio/0.2.22/tokio/sync/broadcast/index.html
+
+An inventory advertisement is an `(InventoryHash, SocketAddr)` pair.  The
+stream hook should check whether an incoming message is an `inv` message with
+only a small number (e.g., 1) inventory entries.  If so, it should extract the
+hash for each item and send it through the channel.  Otherwise, it should
+ignore the message contents.  Why?  Because `inv` messages are also sent in
+response to queries, such as when we request subsequent block hashes, and in
+that case we want to assume that the inventory is generally available rather
+than restricting downloads to a single peer.  However, items are usually
+gossiped individually (or potentially in small chunks; `zcashd` has an internal
+`inv` buffer subject to race conditions), so choosing a small bound such as 1
+is likely to work as a heuristic for when we should assume that advertised
+inventory is not yet generally available.
+
+## Inventory Registry
+
+The peer set's `poll_ready` implementation should extract all available
+`(InventoryHash, SocketAddr)` pairs from the channel, and log a warning event
+if the receiver is lagging.  The channel should be configured with a generous
+buffer size (such as 100) so that this is unlikely to happen in normal
+circumstances.  These pairs should be fed into an `InventoryRegistry` structure
+along these lines:
+
+```rust
+struct InventoryRegistry(HashMap<InventoryHash, HashSet<SocketAddr>>);
+
+impl InventoryRegistry {
+    pub fn register(&mut self, item: InventoryHash, addr: SocketAddr) {
+	self.0.entry(item).or_insert(HashSet::new).insert(addr);
+    }
+
+    pub fn peers(&self, item: InventoryHash) -> Option<&HashSet<SocketAddr>> {
+	self.0.get(item)
+    }
+}
+```
+
+This API does not provide a way to remove inventory advertisements.  Instead,
+the peer set should maintain a `tokio::time::Interval` with some interval
+parameter, and check in `poll_ready` whether the timer has elapsed.  If so, it
+should use `std::mem::take` to extract and drop the current inventory registry.
+This is much simpler than trying to manage the lifetime of individual
+advertisements.  The downside is that if we drop the inventory registry before
+we look up some data, we may not be able to find it until it's widely
+distributed.  To minimize the chance of this happening, the timer check should
+be done **before** processing new channel entries, and the timeout could be
+chosen to be coprime to the block interval (e.g., 79 seconds).
```

Comment:
I think we could use `im` here to avoid dropping recently added entries without increasing the complexity, the difference would be that we clone the data structure on the timer and then if we already have a clone when we try to clone we first remove the overlap between the two sets then replace the old clone with a clone of the leftover difference between the sets.

Reply:
Sorry, I don't think I understand the update procedure you're describing.

Reply:
Branching off of the discord discussion about this suggestion, changed in https://github.com/ZcashFoundation/zebra/pull/952/commits/df93d37ab917eb63b81334e6b8b272350946b349 to add a generational pruning mechanism.